### PR TITLE
dev/financial#180 Switch to expecting total_amount to be tax inclusive (always)

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
My expectation is this will fix for @karing but cause a test to fail - which should help find the
inconsistency

My feeling looking at this is that perhaps historically (at the bao level possibly) it was treated as inclusive if tax_amount is provided & not otherwise.

If my theory is correct I think we could go back to doing that but throw a deprecation notice is tax_amount is provided

note that total_amount is ignored when line items are provided

I'm open to considering a new 'total_amount_inclusive' param - which would be either /or with total_amount (probably throw an exception if both are provided) - that would be in master I guess & I'm not sure whether it's manageable at the moment (ie the reason for wanting to pass in inclusive is to get the rounding right but I don't know how ready the code is for that)

Overview
----------------------------------------
dev/financial#180 Switch to expecting total_amount to be tax inclusive (always)

Before
----------------------------------------
total_amount is treated as tax exclusive when tax_amount is passed in

After
----------------------------------------
- total_amount is IGNORED in favour of using the line items if they are passed in (unchanged)
- if passed in then it is always treated as inclusive
 
In all cases tax_amount is calculated within the order api, based on the financial types & the total amounts passed in (at a line item level or at the api level)

Technical Details
----------------------------------------
This handling has a confusing history and to be honest I thought I'd figured out what would be a no-change behaviour until @KarinG started sleuthing.

I guess what we need to do is 

1) agree the above
2) update docs to make it clear
3) email the dev list

Comments
----------------------------------------

### CODE REVIEW

**Comment by KarinG** [comment]
Hi @eileenmcnaughton -> please note I'm **not trying to pass in 'a tax exclusive amount':** I would never do that! :-)

i) after we (MattWire and I) replaced Transact API with Order API in webform civicrm module -> we have always passed in total_amount for the Order (tax-inclusive) AND with tax_amount in the params. The test that fails now has been in there since we refactored that. In my screenshots (on the issue) $10 is the total_amount of which $0.25 is tax and yes we are providing tax_amount in the params to Order API as well. 

Expected (and ✅ <= 5.40: passing in $10 for total_amount, $0.25 for tax - should create a Contribution that has total_amount = $10.

What is happening ❌ in 5.41rc + dev-master -> it creates a Contribution that has total_amount = $9.75

ii) if we were to introduce a new param -> perhaps total_amount should just remain total_amount - and pre-tax can be **subtotal**. That's what we call it in North America -> here's me ordering two more Buff masks for back to school + stocking up on filter replacements!

image

**Comment by eileenmcnaughton** [comment]
@KarinG this actually passes. I have a feeling I fixed some tests to do valid financials & removed the scenario & I guess that might be why the tax_amount being passed scenario is gone.

I'd be ok to clean this up & merge it (add  a test, fix comments) - I think & then maybe outline it in the dev digest so if anyone has any edge cases where they pass in 'tax_amount' and have different expectations we can figure out how to handle

**Comment by eileenmcnaughton** [comment]
Oh - we cross posted - reading

**Comment by KarinG** [comment]
Oh I think we crossed comments ->

**Comment by KarinG** [comment]
😂

**Comment by eileenmcnaughton** [comment]
@KarinG OK and where you have 

total_amount $10 and tax_amount $.25 - you 

1) are using a financial type with a tax rate of .025%
2) expect the  outcome to be total_amount = $10 and on the line item it would be 9.75 with a tax amount of .25

**Comment by KarinG** [comment]
Will be right back - making coffee - I know it’s evening but that’s ok!

**Comment by KarinG** [comment]
Ok -> 

2. expect the outcome to be total_amount = $10 and on the line item it would be 9.75 with a tax amount of .25

**Comment by eileenmcnaughton** [comment]
@KarinG - can you test / analyse this PR now?

**Comment by eileenmcnaughton** [comment]
(note the test case uses 5% because that was already configured)

**Comment by KarinG** [comment]
I'll test! Can you please edit the title though? We're not switching to expecting Tax exclusive.

**Comment by eileenmcnaughton** [comment]
@KarinG done

File: api/v3/Order.php [hunk 1]
```diff
@@ -102,9 +102,9 @@ function civicrm_api3_order_create(array $params): array {
     $order->setPriceSetToDefault('contribution');
     $order->setLineItem([
       // Historically total_amount in this case could be tax
-      // inclusive if tax is also supplied.
+      // inclusive if tax is also supplied - as of 5.41 it always is.
       // This is inconsistent with the contribution api....
-      'line_total' => ((float) $params['total_amount'] - (float) ($params['tax_amount'] ?? 0)),
+      'line_total_inclusive' => (float) $params['total_amount'],
       'financial_type_id' => (int) $params['financial_type_id'],
```
Comment (KarinG) [inline api/v3/Order.php:108]: I'm going to check this first.

Comment (eileenmcnaughton) [reply api/v3/Order.php:108]: @KarinG ok - I've updated it slightly although that line should be the same

Comment (KarinG) [reply api/v3/Order.php:108]: Changing this line alone -> fixes the regression. I now get $10 - test passes.

File: tests/phpunit/api/v3/OrderTest.php [hunk 2]
```diff
@@ -488,6 +488,24 @@ public function testDeleteOrder(): void {
...
+    $order = $this->callAPISuccess('Order', 'create', [
+      'total_amount' => 10,
+      'financial_type_id' => 'Test taxable financial Type',
+      'contact_id' => $this->individualCreate(),
+      'sequential' => 1,
+    ])['values'][0];
+    $this->assertEquals(10, $order['total_amount']);
+    $this->assertEquals(0.47619047619048, $order['tax_amount']);
+  }
+
```
Comment (KarinG) [inline tests/phpunit/api/v3/OrderTest.php:508]: Ah I see -> so if we're passing in $10 and $0.50 e.g. The $10 and $0.50

But if the financial type is of a type that has Sales Tax associated with it then we need to API to calculate sales tax for us. And that then becomes 0.47619047619048 ✅

Comment (eileenmcnaughton) [reply tests/phpunit/api/v3/OrderTest.php:508]: @KarinG right - the assumption is that the tax amount can always be calculated from the financial type & the function calling order api just passes the total amount (inclusive!) - so tax_amount is actually a deprecated parameter for the order api

File: CRM/Financial/BAO/Order.php [hunk 3]
```diff
@@ -1122,4 +1123,20 @@ protected function fillDefaultContributionLine(array &$lineItem): void {
...
+   * @param $inclusiveAmount
+   *
+   * @return array
+   */
+  protected function alterLineItemForTax($taxRate, array $lineItem, $inclusiveAmount): array {
+    // Total is tax inclusive.
+    $lineItem['tax_amount'] = ($taxRate / 100) * $inclusiveAmount / (1 + ($taxRate / 100));
+    $lineItem['line_total'] = $lineItem['unit_price'] = $inclusiveAmount - $lineItem['tax_amount'];
+    return $lineItem;
+  }
```
Comment (KarinG) [inline CRM/Financial/BAO/Order.php:1140]: Will wait for the Jenkins to run first then pull this in.

File: tests/phpunit/api/v3/OrderTest.php (hunk 2 above)
Comment (KarinG) [reply tests/phpunit/api/v3/OrderTest.php:508]: That would be a change in what the API has been doing through 5.40 though. We would have to (and could) rewrite the way we have been using Order API in the workflow. But it would be a change - (that others may hit as well)

End result for our (complicated scenario - mix and match) test is actually adding line_items ad-hoc:

Your edit to this single line -> 
'line_total' => (float) $params['total_amount'],

Makes the API work as it did before ✅

image

Why change more things? I'm a bit worried about calculating things that don't need to be calculated - if I'm passing sales tax into the API I would expect it to return as is. 

I've always looked at Order API as creating the Order (Contribution ID) so that we can then attempt the Payment and then attach the Payment (processor just wants total_amount). Then we figure out the line items/details and write them to the record and decide who needs registrations and/or membership edits. Now some of that logic appears to be shifting to the Order API. 

E.g. note that in the example screenshot -> Main Financial Type (at the Contribution level) is Donation - so 0% Sales Tax. Yet it contains a line_item that are subject to Sales Tax. That's a real scenario -> making a Donation so that Contribution at the highest level is deductible for tax purposes - but adding a purchase to transaction.

Comment (KarinG) [reply tests/phpunit/api/v3/OrderTest.php:508]: I'll be checking in with J tomorrow. I'll ask him to look at our logic/use of Order API as well -> this is what Matt and I ended up with as a straight up replacement for Transact API -> 

```
  function wf_civicrm_api3_contribution_transact($params) {
    // Start with the same parameters as Contribution.transact.
    $params['contribution_status_id'] = 'Pending';
    if (!isset($params['invoice_id']) && !isset($params['invoiceID'])) {
      // Set an invoice_id here if you have not already done so.
      // Potentially Order api should do this https://lab.civicrm.org/dev/financial/issues/78
      $params['invoice_id'] = md5(uniqid(rand(), TRUE));
    }
    if (isset($params['invoice_id']) && !isset($params['invoiceID'])) {
      // This would be required prior to https://lab.civicrm.org/dev/financial/issues/77
      $params['invoiceID'] = $params['invoice_id'];
    }
    elseif (!isset($params['invoice_id']) && isset($params['invoiceID'])) {
      $params['invoice_id'] = $params['invoiceID'];
    }

    $order = civicrm_api3('Order', 'create', $params);
    try {
      // Use the Payment Processor to attempt to take the actual payment. You may
      // pass in other params here, too.
      $propertyBag = new \Civi\Payment\PropertyBag();
      $propertyBag->mergeLegacyInputParams($params);
      $propertyBag->setContributionID($order['id']);
      // @fixme: Class \Civi\Payment\CRM_Utils_Money not found (fixed in 5.28 via https://github.com/civicrm/civicrm-core/pull/17505
      //   But note we still return a localized format
      // $propertyBag->setAmount($params['total_amount']);

      // We need the payment params as an array to pass to PaymentProcessor.Pay API3.
      // If https://github.com/civicrm/civicrm-core/pull/17507 was merged we wouldn't need this hack
      //   using ReflectionClass to access the array.
      $reflectionClass = new \ReflectionClass($propertyBag);
      $reflectionProperty = $reflectionClass->getProperty('props');
      $reflectionProperty->setAccessible(TRUE);
      $payParams = $reflectionProperty->getValue($propertyBag)['default'];

      // propertyBag has 'contributionID', we need 'contribution_id'.
      $payParams['contribution_id'] = $propertyBag->getContributionID();
      // propertyBag has 'currency', iATS uses `currencyID` until https://github.com/iATSPayments/com.iatspayments.civicrm/pull/350 is merged
      $payParams['currencyID'] = $propertyBag->getCurrency();
      $payParams['amount'] = $params['total_amount'];
      $payResult = reset(civicrm_api3('PaymentProcessor', 'pay', $payParams)['values']);

      // webform_civicrm sends out receipts using Contribution.send_confirmation API if the contribution page is has is_email_receipt = TRUE.
      // We allow this to be overridden here but default to FALSE.
      $params['is_email_receipt'] = $params['is_email_receipt'] ?? FALSE;

      // Assuming the payment was taken, record it which will mark the Contribution
      // as Completed and update related entities.
      civicrm_api3('Payment', 'create', [
        'contribution_id' => $order['id'],
        'total_amount' => $payParams['amount'],
        'payment_instrument_id' => $order['values'][$order['id']]['payment_instrument_id'],
        // If there is a processor, provide it:
        'payment_processor_id' => $payParams['payment_processor_id'],
        'is_send_contribution_notification' => $params['is_email_receipt'],
        'trxn_id' => $payResult['trxn_id'] ?? NULL,
      ]);
    }
    catch (\Exception $e) {
      return ['error_message' => $e->getMessage()];
    }

    $contribution = civicrm_api3('Contribution', 'getsingle', ['id' => $order['id']]);
    // Transact API works a bit differently because we are actually adding the payment and updating the contribution status.
    // A "normal" transaction (eg. using doPayment() or just PaymentProcessor.pay) would not update the contribution and would
    // rely on the caller to do so. The caller relies on the `payment_status_id` (Pending or Completed) to know whether the payment
    // was successful or not.
    $contribution['payment_status_id'] = $contribution['contribution_status_id'];
    return $contribution;
  }
```

Comment (eileenmcnaughton) [reply tests/phpunit/api/v3/OrderTest.php:508]: @KarinG the main reason for changing things was that there were various places where financials were winding up incorrect & the schema is supposed to be that the financial type id dictates the tax rate (& I don't think it's reliable without). Note this PR does not make that change

Regarding your scenario - aren't you talking about a case where you are passing in more than one line items & wouldn't the tax-attractiving line item have a tax-attracting financial type? It is per line item

Comment (KarinG) [reply tests/phpunit/api/v3/OrderTest.php:508]: In the example test/scenario -> the Order is created with $10 and $0.25 tax (on the Contribution level) and a non-tax-attracting financial type at the highest level - we've never passed in any line_items at the stage of either Transact API or now Order API. We could start doing that -> we are doing our own tally of line items to get to the correct amounts in the first place $10 and $0.25 (obviously).

Comment (eileenmcnaughton) [reply tests/phpunit/api/v3/OrderTest.php:508]: @KarinG yeah - if you want order api to work with multiple line items the expectation is that you pass in the line items - passing in total_amount is really only for single line items

Comment (eileenmcnaughton) [reply tests/phpunit/api/v3/OrderTest.php:508]: Note I'm not sure which version but I have made it smarter about requiring less keys - this is from 5.41 tests but I think it figures out some of those values if not passed 

```
 $params = [
      'contact_id' => $this->_individualId,
      'receive_date' => '2018-01-01',
      'total_amount' => 50,
      'financial_type_id' => $this->_financialTypeId,
      'contribution_status_id' => 'Pending',
      'line_items' => [
        0 => [
          'line_item' => [
            '0' => [
              'price_field_id' => 1,
              'price_field_value_id' => 1,
              'label' => 'Test 1',
              'field_title' => 'Test 1',
              'qty' => 1,
              'unit_price' => 40,
              'line_total' => 40,
              'financial_type_id' => 1,
              'entity_table' => 'civicrm_contribution',
            ],
          ],
        ],
      ],
    ];
```

File: CRM/Financial/BAO/Order.php (hunk 3 above)
Comment (eileenmcnaughton) [reply CRM/Financial/BAO/Order.php:1140]: Ah - those are those test fails that I thought would happen - this gets me back to 'this api has always behaved differently depending on ... something'

@KarinG a couple of notes to your usage
1) we need to make sure the rc doesn't break your usage & find way to do that
2) against master we would ideally get your code to call the order api correctly (ie if there is more than one line item then pass them in OR (less good) just switch to doing exactly the same code but call Contribution.create with contribution_status_id = Pending which I think is functionally equivalent the way you are using the code (given you then delete & recreate the line items). I'd like to think that would be an interim thing & I'm totally happy to work with you on getting to 'full line item'


```
Stacktrace
api_v3_ContributionTest::testCheckTaxAmount with data set #0 ('.')
Failed asserting that '4.76' matches expected 5.

/home/jenkins/bknix-dfl/build/core-21187-75kjs/web/sites/all/modules/civicrm/tests/phpunit/api/v3/ContributionTest.php:2284
/home/jenkins/bknix-dfl/build/core-21187-75kjs/web/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php:258
/home/jenkins/bknix-dfl/extern/phpunit8/phpunit8.phar:671

```

Comment (eileenmcnaughton) [reply CRM/Financial/BAO/Order.php:1147]: @KarinG I think https://github.com/civicrm/civicrm-core/pull/21187/commits/0f4482bc116fb83e1985352e6c05ec28e03bd989 will allow your override tax amount to keep working until we have time to address it - let's see how tests go. 

Also your work clarifying the contribution_recur was useful in knowing how to deal with the test fail :-)

**Comment by KarinG** [comment]
Thanks @eileenmcnaughton 

My summary:

**Before ->** Order API -> line items were essentially optional to pass in. One could create a Contribution with Financial Type (on the Contribution) and Tax amount (on the Contribution) - and passing in those two would do exactly that -> create a Contribution with the correct values at the Contribution level.

**Trying to move towards ->** line items are mandatory -> passing in Total Amount (on the Contribution) and Tax Amount (on the Contribution) should be completely deprecated as all of these can be calculated from the line item Level. 

I'm worried about you editing unit tests in this PR. You're always miles ahead of me in thinking. 

To mitigate/address the regression - we could we just do this -> (this was your initial commit on this PR and Jenkins was happy and I confirmed this solves the regression:

`'line_total' => (float) $params['total_amount'],`

And then work on making line items required?

**Comment by eileenmcnaughton** [comment]
@KarinG so this PR doesn't make line items required. The expectation is that line items would always be passed in unless there is only one - that hasn't changed (this discussion obviously I learnt you were doing something unexpected :-)). I did add to this PR to try to make it handle the situation you described where you pass in an override-tax-amount because you not providing the api the information to calculate the tax amount (and proposed we work together on a longer term fix)

Making this change `'line_total' => (float) $params['total_amount'],` turned out to be wrong - because the line items don't store the inclusive amount in `line_total`

Regarding the test changes -2 of those tests had recently been changed from calling Contribution.create to Order.create so I think it's OK to update them. For repeattransaction I *think* the behaviour change from this PR made something that was not correct correct - I would definitely like to confirm that.


I'm increasingly sure the old behaviour was 

- if you called the order api without specifying line items and with total_amount AND tax_amount, total_amount was treated as inclusive
- if you called the order api  without specifying line items and with only total_amount, total_amount was treated as exclusive
- ditto for contribution create but presence or otherwise of line items were irrelevant

If I'm correct then the technical term for the behaviour is `batshit crazy`

It would be good to confirm the above against something a bit older (eg. 5.33)

I think the correct behaviour for order api is that it should always be inclusive (I had started to resolve it should always be exclusive but I'm mostly concerned with that always word)

Dang I really need your brain helping me make sense of this

**Comment by KarinG** [comment]
I need some uninterrupted time on this (tomorrow) and will run a bunch of tests/scenarios.

First and foremost, I agree the RC failed for us b/c we've done line items ad-hoc (for 10y I think - since Coleman first added line items for webform civicrm) - and it could well be time to change that workflow.

But the real issue we should be trying to solve so that it will be clear forever and ever -> 
is that currently total_amount can have two very different meanings:
total_amount was treated as inclusive vs total_amount was treated as exclusive.

I should not be able to pass in total_amount = $10 to an API which then returns total_amount = $9.75 (because Order API is identifying a different context) - it's the same variable - the work arounds may well work now but we need to fix the definition and/or introduce subtotal_amount to clearly differentiate between total_amount (including sales taxes if any apply) and subtotal_amount (is always before sales tax)

Then the brain teasers where we're recalculating total_amount based on context will be a lot more readable.

More tomorrow!

**Comment by eileenmcnaughton** [comment]
@KarinG OK - I think we are on the same thought path here & I really appreciate being ablt to work through it with you. We might need to break it into 2 parts

1) a short term fix for the rc
2) the right fix & however we migrate to that

File: CRM/Financial/BAO/Order.php [hunk 4]
```diff
@@ -940,6 +940,16 @@ public function setLineItem(array $lineItem, $index): void {
     if ($this->getOverrideTotalAmount()) {
       $this->addTotalsToLineBasedOnOverrideTotal((int) $lineItem['financial_type_id'], $lineItem);
     }
+    elseif (isset($lineItem['line_total_inclusive'])) {
+      if (isset($lineItem['tax_amount_override'])) {
+        // @todo deprecate this - work with Karin to make sure she is not using order api in
+        // a way that deprecating this causes problems.
+        $lineItem['line_total'] = $lineItem['line_total_inclusive'] = $lineItem['tax_amount'];
```
Comment (jitendrapurohit) [inline CRM/Financial/BAO/Order.php:947]: @eileenmcnaughton should this be `$lineItem['tax_amount_override']` on the right? Currently if i submit a webform with a taxable financial type, it creates a contribution with amount = $0. Probably because line_total is not set in this step. A quick way to replicate is execute this API -

```
$result = civicrm_api3('Order', 'create', [
  'contact_id' => 202,
  'financial_type_id' => "Member Dues",
  'total_amount' => 220,
  'tax_amount' => 20,
]);
```
where member dues is a taxable ft of 10%.

**Comment by KarinG** [comment]
Ok starting to look at this now 🎉 I will be posting thoughts and updates here.

**Comment by KarinG** [comment]
1. public function testOnePageDonation() ✅ (financial type -Donation- is not subject to sales tax) passes w/RC

2. Ok reproduced locally ->   <version_no>5.41.beta1</version_no>
image

**Comment by KarinG** [comment]
3. Let's check to see if all is well if we restructure webform civicrm to pass on all line items - as we definitively know them and they are all tallied up properly before submitting the form:

image

**Comment by KarinG** [comment]
Testing this PR with the small edit we noted this morning - getting there! 🙂

But before looking at the API - thinking about what we need it to look like:

- if lineitems (even if there is only 1) -> always do the math (i.e. recalculate $params['total_amount'] and $params['tax_amount] passed into Order API). Spit out a deprecation notice. No need to put these in as we'll not be using them to construct the Contribution record.

Else
- if there are no lineitems -> puzzle time -> we need to put together a line_item based on total_amount + financial_type_id.

**This is a puzzle regression** -> the puzzle for our WFC tests goes wrong b/c we presented everything at a Contribution level:

$10 total_amount
$0.25 tax_amount
main financial type (on the contribution record) -> Donation (so no sales tax). 

And it's **_not possible_** to make a coherent line item out of these three inputs
The puzzle went like this -> $10 - $0.25 -> $9.75 -> then add 0% = $9.75 [new total_amount]

**Comment by KarinG** [comment]
Conclusion re: the regression:

Previously (<= 5.40) it was possible to ask Order API to create incoherent line items - as of 5.41rc it's not 🙂 Actually -> it's more like we are/were just wanting to process $10 rather than to create any line items at all [we are/were using it as a replacement of Transaction API]

What to do... 

Some testing on the passing in no line item/puzzle scenario - but with coherent data:

Passing in:
$10 total_amount
$0.50 tax_amount
financial type = Purchase 5 (i.e. 5% sales tax)

$Order out -> 
$9.975 total_amount ❌
$0.50 tax_amount

image




1. short term fix for the rc ->

**Comment by KarinG** [comment]
Intermediate conclusion there is another Regression in the 5.41rc (this is all still before testing this PR).

image

image

**Comment by KarinG** [comment]
- If we'd say let's not fix the regression b/c it is impossible to come up with proper lineitems for that scenario (passing in 3 params that are incoherent) -> I'd understand. 

- If we'd say let's fix the regression ($order used to return $10 in this scenario and now it returns $9.75):

image

This is not about tax inclusive or tax exclusive - It's also not about any overrides - it's just not possible to create a coherent lineitem.

The only solution for the regression would be if Order API would not create any lineitems at all [based on params input]. That's the use case of the Order API that's seeing the regression. Those that just want to block in a Contribution record with main Financial Type (on the Contribution level), Total Amount (on the Contribution level) and Tax Amount (on the Contribution level).  

Ok that's my summary thus far. I'll take a break and will check back in later! I still have some more testing to do but had to reason through the logic first.

**Comment by eileenmcnaughton** [comment]
@KarinG - @jitendrapurohit was right about an issue - I'm just digging into that & have spotted a related issue & coming up with an altered version - will put more in a mo

**Comment by KarinG** [comment]
Yes - we discussed it this morning. I'm now thinking we need to do less. 

Sorry - had to take a break and am now in a work meeting with one of my team.

**Comment by eileenmcnaughton** [comment]
@KarinG I found the original sin - it's this line https://github.com/civicrm/civicrm-core/commit/e967ce8fe2b58b94e2163dde395542e55599da13#diff-a16d4d7449cf5f3a0616d1d282a32f27ab6d3f7d2726d076c02ad1d4d655af41R512 is assuming the total_amount (already set on line_total) to be exclusive - but it's in fact inclusive. I think this led to other stuff tho so I don't think just fixing it is enough - but I'm close to having what I think is reasonable fix for the regression

**Comment by KarinG** [comment]
@eileenmcnaughton -> I've come to the conclusion (see my comments above) that this is not a tax inclusive or tax exclusive issue. It's a set of inputs for which it's 'impossible to create a coherent line item' issue.

**Comment by eileenmcnaughton** [comment]
@KarinG I think the update I pushed up will fix the regression (and I'm only focussing on the regression for now).

To recap - the issue that drove a lot of this effort to ensure financials were valid was a bug identified by @christianwach whereby orders with incorrectly calculated tax_amounts were changing when you edited them & this happens on 5.33 in my testing in pretty much all of the ways you can create a 'simple' contribution with tax.

The 4 methods I've focussed on are 

- contribution.create with tax passed in
- contribution.create without tax passed in
- order.create with tax passed in
- order.create without tax passed in

In all cases I used 5% tax on financial type campaign contribution with total_amount = 105 and tax_amount = 5 (where used)

Without this PR the situation for the above 4 uses is bad, primarily due to the bug above but I think that not spotting that led to other issues.

image


By contrast in 5.33 here is what we see - the fact tax is NOT being calculated on the line items is the key cause of the edit bug

image

**Comment by eileenmcnaughton** [comment]
@KarinG so in summary - I think this fixes it so you shouldn't need to change your code for it to work with 5.41 and if tests pass then I'm happy with it at this point

I have a  suspicion tests won't all pass because I think a lot of them were written / updated based on that original sin - but I think that is OK to fix.

**Comment by KarinG** [comment]
Sorry if I've not made this clear - but it's not an inclusive vs exclusive issue. I'm worried about the whack-a-mole here - lots of files/unit tests updated. It's impossible to do the math to coherent line item with total_amount = $10, tax_amount = $0.25 and financial type = Donation. The only solution in this case is to not create a line item.

**Comment by eileenmcnaughton** [comment]
Regarding the longer term order api - my expectation is there are 2 valid ways to call it

1) you are creating an order with a single line item. Required fields are financial_type_id, contact_id and total_amount
2) you are creating an order with one or more line items. Required fields are 'contact_id' and 'line_items'. 

In both cases 'tax_amount' would be deprecated and eventually completely ignored. Passing in both line items & total_amount should also probably be deprecated.

Within line items there is a bit of variation across versions as to the minimum required - I've been trying to reduce it down to what cannot be intuited from other values (e.g if you don't pass in qty then we could probably assume 1 & if you don't pass in unit_price we could assume it === line_total) - but that is probably out of scope for this conversation

**Comment by eileenmcnaughton** [comment]
@KarinG did you re-test with the latest version - I think it addresses your original issue.

**Comment by KarinG** [comment]
No - I've not. It does way more than just fix the regression - which worries me. 

The regression:

a) I don't think should be fixed but rather return an error 'incompatible set of input parameter: order API can not create a coherent line-item based on the input $params'. 

b) it only 'worked' for us before b/c total_amount passed in was the same as total_amount returned by Order API.

**Comment by KarinG** [comment]
I am planning on running tests tomorrow.

**Comment by eileenmcnaughton** [comment]
@KarinG no it doesn't actually go further than fix the regression - it fixes the code to do the same thing that it did in 5.40 so that is

1) keeps the total amount as passed in as the tax inclusive amount
2) keeps the tax amount if passed in as the tax_amount

and it does ^^ for both contribution and order now

The test changes are

**straight forward **
1) add tests 
2) update 3 tests - but note that all 3 of these tests changed in some way after the original sin happened. In all cases the goal was to fix aspects of them that were doing set up with incorrect financials - but because that was done after *the mistake* they were locking it in

**Comment by eileenmcnaughton** [comment]
Hmm - the problem with talking on PRs like this is that when we close them (like I'm going to now in favour of #21230 ) the discussion slips away - still I think we captured a lot of it in gitlab & some of it even sunk into my brain (and a lot of it is based on first cut comprehension of the issue)
